clients/csrf/transport.go

```diff
@@ -7,48 +7,23 @@ import (
 	"github.com/cloudfoundry-incubator/multiapps-cli-plugin/log"
 )
 
-type Csrf struct {
-	Header              string
-	Token               string
-	IsInitialized       bool
-	NonProtectedMethods map[string]bool
-}
-
-type Cookies struct {
-	Cookies []*http.Cookie
-}
-
 type Transport struct {
-	OriginalTransport http.RoundTripper
-	Csrf              *Csrf
-	Cookies           *Cookies
+	Delegate http.RoundTripper
+	Csrf     *Csrf
 }
 
-func (t Transport) RoundTrip(req *http.Request) (*http.Response, error) {
+func (t *Transport) RoundTrip(req *http.Request) (*http.Response, error) {
 	reqCopy := req.Clone(req.Context())
 
-	csrfTokenManager := NewDefaultCsrfTokenManager(&t, reqCopy)
-	err := csrfTokenManager.updateToken()
+	csrfTokenManager := NewDefaultCsrfTokenManager(t)
+	err := csrfTokenManager.updateToken(reqCopy)
 	if err != nil {
 		return nil, err
 	}
 
-	log.Tracef("Sending a request with CSRF '%s : %s'\n", reqCopy.Header.Get("X-Csrf-Header"), reqCopy.Header.Get("X-Csrf-Token"))
+	log.Tracef("Sending a request with CSRF %s\n", reqCopy.Header.Get("X-Csrf-Token"))
 	log.Tracef("Cookies used are: %s\n", prettyPrintCookies(reqCopy.Cookies()))
-	res, err := t.OriginalTransport.RoundTrip(reqCopy)
-	if err != nil {
-		return nil, err
-	}
-	tokenWasRefreshed, err := csrfTokenManager.refreshTokenIfNeeded(res)
-	if err != nil {
-		return nil, err
-	}
-
-	if tokenWasRefreshed {
-		log.Tracef("Response code '%d' from bad token. Must Retry.\n", res.StatusCode)
-		return nil, &ForbiddenError{}
-	}
-	return res, err
+	return t.Delegate.RoundTrip(reqCopy)
```

Comment:
I think this code handled the case when DS instance is restarted and the token and session are lost.
Now this case is not handled.
The DS will return a header:
"X-CSRF-TOKEN: Required", when Invalid/Missing Csrf exception occurs. We can check this header and re-init the token and re-execute the request.